### PR TITLE
[Rest Api Compatibility] transformations for keys in match and length

### PR DESCRIPTION
**Pull request by pgomulka**
Additional yml tests transformations to modify a key in a match assertion and a key in a
length assertion. 
The testing is done with the original and expected transformed files. This should make it a little more easy to write new tests and add new transformations.

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/rest/transform/match/ReplaceMatch.java [hunk 1]
```diff
@@ -21,18 +21,21 @@
  */
 public class ReplaceMatch implements RestTestTransformByParentObject {
     private final String replaceKey;
+    private final String newKeyName;
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/rest/transform/match/ReplaceMatch.java:24]: This needs to be exposed as an `@Input`. Adding a public getter for this field would also remove the need to override it in `ReplaceKeyInMatch`.

**Review by mark-vieira — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java [hunk 2]
```diff
@@ -105,9 +107,40 @@ public void replaceMatch(String subKey, Object value) {
...
+    }
+
+    /**
+     * A transformation to replace the key in a length assertion.
+     * @see ReplaceLength
+     * @param oldKeyName   the key name directly under length to replace.
+     * @param newKeyName   the new key name directly under length.
+     */
+    public void replaceKeyInLength(String oldKeyName, String newKeyName) {
+        transformations.add(new ReplaceLength(oldKeyName, newKeyName, null));
```
Comment (joegallo) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java:128]: Minor nit: what do you think about renaming `ReplaceLength` to `ReplaceKeyInLength`? The rest already match the `fooBar`/`FooBar` convention, but this one doesn't.

File: buildSrc/src/test/java/org/elasticsearch/gradle/internal/test/rest/transform/length/ReplaceLengthTests.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    public void testLengthKeyChange() throws Exception {
+
+        String test_original = "/rest/transform/length/length_replace_original.yml";
+        List<ObjectNode> tests = getTests(test_original);
+
+        String test_transformed = "/rest/transform/length/length_replace_transformed.yml";
+        List<ObjectNode> expectedTransformation = getTests(test_transformed);
+
+        List<ObjectNode> transformedTests = transformTests(
+            new LinkedList<>(tests),
```
Comment (joegallo) [inline buildSrc/src/test/java/org/elasticsearch/gradle/internal/test/rest/transform/length/ReplaceLengthTests.java:32]: I think `transformTests` already does a defensive copy of the list of tests internally (though, to be completely honest, I'm not especially sure why it bothers).

**Review by jakelandis — CHANGES_REQUESTED** [review]
A few suggestions on class organization.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java (hunk 2 above)
Comment (joegallo) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java]: 563e86f

File: buildSrc/src/test/java/org/elasticsearch/gradle/internal/test/rest/transform/length/ReplaceLengthTests.java (hunk 3 above)
Comment (joegallo) [reply buildSrc/src/test/java/org/elasticsearch/gradle/internal/test/rest/transform/length/ReplaceLengthTests.java]: ea19773

**Review by mark-vieira — APPROVED** [review]
LGTM

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java [hunk 4]
```diff
@@ -105,9 +107,40 @@ public void replaceMatch(String subKey, Object value) {
...
     }
 
+    /**
+     * Replaces the values of a match assertion for the given REST test. For example "match":{"_type": "foo"} to "match":{"_type": "bar"}
+     *
+     * @param subKey   the key name directly under match to replace. For example "_type"
+     * @param value    the value used in the replacement. For example "bar"
+     * @param testName the testName to apply replacement
+     */
+    public void replaceMatch(String subKey, Object value, String testName) {
```
Comment (jakelandis) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java:117]: nit: can you rename this (and the associated class to replaceValueInMatch ? to help distinguish the difference between this and replaceKeyInMatch.

**Review by jakelandis — APPROVED** [review]
Changes look good! (one minor, non blocking nitpick request)

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java (hunk 4 above)
Comment (joegallo) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java]: I think f64538ab84316a01ea4f5dc406b1510ae06b7fc2 should cover what you're asking for here.

Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java]: @joegallo - thanks! LGTM

**Pull request merged**